{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 231115\n",
    "# script for github"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from: https://colab.research.google.com/github/zaneveld/full_spectrum_bioinformatics/blob/master/content/06_biological_sequences/reading_and_writing_fasta_files.ipynb#scrollTo=ZkpJZr86_9Jz\n",
    "def fastaDictionary(FastaFile):\n",
    "    fasta = open(FastaFile, \"r\")\n",
    "    \"\"\"Return a dict of {id:gene_seq} pairs based on the sequences in the input FASTA file\n",
    "    input_file -- a file handle for an input fasta file\n",
    "    \"\"\"\n",
    "    parsed_seqs = {}\n",
    "    curr_seq_id = None\n",
    "    curr_seq = []\n",
    "\n",
    "    for line in fasta:\n",
    "        line = line.strip()\n",
    "        \n",
    "        #set up the dictionary\n",
    "        if line.startswith(\">\"):\n",
    "            # before going on to the next sequence in the file, join the final sequences together in the dictionary\n",
    "            if curr_seq_id is not None:\n",
    "                parsed_seqs[curr_seq_id] = ''.join(curr_seq)\n",
    "                \n",
    "            # rename/update/ curr_seq_id if there is a new line with \">\"\n",
    "            curr_seq_id = line[1:]\n",
    "            # reset the sequence list to join\n",
    "            curr_seq = []\n",
    "            continue\n",
    "        \n",
    "        #add the sequence to the list\n",
    "        curr_seq.append(line)\n",
    "\n",
    "    #Add the final sequence to the dict\n",
    "    parsed_seqs[curr_seq_id] = ''.join(curr_seq)\n",
    "    return(parsed_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract sequences from Chi, Sal, and Tar\n",
    "def extract_ChiSalTar(FastaFile, targetGene, outputFasta):\n",
    "    fastaDict = fastaDictionary(FastaFile)\n",
    "    if targetGene in fastaDict.keys():\n",
    "        #print(targetGene)\n",
    "        #print(fastaDict[targetGene])\n",
    "        newFasta = open(outputFasta, \"a+\")\n",
    "        newFasta.write(\">\"+targetGene)\n",
    "        newFasta.write(\"\\n\")\n",
    "        newFasta.write(fastaDict[targetGene])\n",
    "        newFasta.write(\"\\n\")\n",
    "        newFasta.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract sequences from Ath, Csa\n",
    "def extract_AthCsa(FastaFile, targetGene, outputFasta):\n",
    "    fastaDict = fastaDictionary(FastaFile)\n",
    "    for gene_id, seq in fastaDict.items():\n",
    "        #print(gene_id)\n",
    "        #print(seq)\n",
    "        geneIDlist = gene_id.strip().split(\" \")\n",
    "        fullName = geneIDlist[0].split(\".\")\n",
    "        transcriptName = fullName[0] + \".\"+ fullName[1]\n",
    "        if targetGene == transcriptName:\n",
    "            newFasta = open(outputFasta, \"a+\")\n",
    "            newFasta.write(\">\"+targetGene)\n",
    "            newFasta.write(\"\\n\")\n",
    "            newFasta.write(seq)\n",
    "            newFasta.write(\"\\n\")\n",
    "            newFasta.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract sequences from Aar, Bra, Cru\n",
    "def extract_AarBraCru(FastaFile, targetGene, outputFasta):\n",
    "    fastaDict = fastaDictionary(FastaFile)\n",
    "    for gene_id, seq in fastaDict.items():\n",
    "        #print(gene_id)\n",
    "        #print(seq)\n",
    "        geneIDlist = gene_id.strip().split(\" \")\n",
    "        fullName = geneIDlist[0]\n",
    "        if targetGene == fullName:\n",
    "            newFasta = open(outputFasta, \"a+\")\n",
    "            newFasta.write(\">\"+targetGene)\n",
    "            newFasta.write(\"\\n\")\n",
    "            newFasta.write(seq)\n",
    "            newFasta.write(\"\\n\")\n",
    "            newFasta.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# list of genes to extract from target list\n",
    "def ATgeneList(ATgeneList_path, OG_path):\n",
    "    \n",
    "    # open inclusive orthogroup file\n",
    "    OGdf = pd.read_csv(OG_path)\n",
    "    OGdf = OGdf.fillna(\"*\")\n",
    "    \n",
    "    # create dictionary from OG list\n",
    "    OG_Dict = {}\n",
    "    for index, row in OGdf.iterrows():\n",
    "        atGene = row[1]\n",
    "        if atGene not in OG_Dict:\n",
    "            OG_Dict[atGene] = []\n",
    "            for i in range(2,10):\n",
    "                if row[i] != \"*\":\n",
    "                    geneList = row[i].split(\", \")\n",
    "                    for gene in geneList:\n",
    "                        OG_Dict[atGene].append(gene)\n",
    "    \n",
    "    # open target gene list\n",
    "    ATgeneList = open(ATgeneList_path, \"r\")\n",
    "    geneOGList = []\n",
    "    for gene in ATgeneList:\n",
    "        gene = gene.strip()\n",
    "        if gene in OG_Dict:\n",
    "            targetGeneList = OG_Dict[gene]\n",
    "            geneOGList.append([gene, targetGeneList])\n",
    "    \n",
    "    return(geneOGList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run script to extract sequences\n",
    "main_path = r\"/path_to_fasta\"\n",
    "# path to protein FASTA files\n",
    "AarFastaFile = main_path + \"/Ae.arabicum_v3.1_proteins_noTE_23160.fasta\"\n",
    "AthFastaFile = main_path + \"/Athaliana_447_Araport11.protein_primaryTranscriptOnly.fa\"\n",
    "BraFastaFile = main_path + \"/BrapaFPsc_277_v1.3.protein_primaryTranscriptOnly.fa\"\n",
    "ChiFastaFile = main_path + \"/Chirsuta.primary.pep.fa\"\n",
    "CruFastaFile = main_path + \"/Crubella_474_v1.1.protein_primaryTranscriptOnly.fa\"\n",
    "CsaFastaFile = main_path + \"/Csativa.primary.pep.mod.fa\"\n",
    "SalFastaFile = main_path + \"/Sal.pep\"\n",
    "TarFastaFile = main_path + \"/Tar_v2_protein_geneID.fa\"\n",
    "\n",
    "# path to inclusive OG file and target gene list (AT gene IDs)\n",
    "inclusiveOG_path = r\"/path_to_inclusiveOG/inclusiveOG_fullSet_230928.csv\"\n",
    "ATtargets_path = r\"/path_to_YABBY_list/AT_YABBY_geneID.txt\"\n",
    "\n",
    "#list of sequences - from a function\n",
    "listOfTargets = ATgeneList(ATtargets_path, inclusiveOG_path)\n",
    "\n",
    "for geneSet in listOfTargets:\n",
    "    targetGene = geneSet[0]\n",
    "    OGgenes = geneSet[1]\n",
    "    \n",
    "    #output file path\n",
    "    ProteinSeqFasta_path = r\"/path_to_out/\"+targetGene+\"_sequence.fa\"    \n",
    "    ALL_YABBY_Fasta_path = r\"/path_to_out/ALL_sequence.fa\"\n",
    "    \n",
    "    for gene in OGgenes:\n",
    "        if gene.startswith(\"Aa\") == True:\n",
    "            extract_AarBraCru(AarFastaFile, gene, ProteinSeqFasta_path)\n",
    "            extract_AarBraCru(AarFastaFile, gene, ALL_YABBY_Fasta_path)\n",
    "        elif gene.startswith(\"AT\") == True:\n",
    "            extract_AthCsa(AthFastaFile, gene, ProteinSeqFasta_path)\n",
    "            extract_AthCsa(AthFastaFile, gene, ALL_YABBY_Fasta_path)\n",
    "        elif gene.startswith(\"Bra\") == True:\n",
    "            extract_AarBraCru(BraFastaFile, gene, ProteinSeqFasta_path)\n",
    "            extract_AarBraCru(BraFastaFile, gene, ALL_YABBY_Fasta_path)\n",
    "        elif gene.startswith(\"CARH\") == True:\n",
    "            extract_ChiSalTar(ChiFastaFile, gene, ProteinSeqFasta_path)\n",
    "            extract_ChiSalTar(ChiFastaFile, gene, ALL_YABBY_Fasta_path)\n",
    "        elif gene.startswith(\"Csa\") == True:\n",
    "            extract_AthCsa(CsaFastaFile, gene, ProteinSeqFasta_path)\n",
    "            extract_AthCsa(CsaFastaFile, gene, ALL_YABBY_Fasta_path)\n",
    "        elif gene.startswith(\"Carub\") == True:\n",
    "            extract_AarBraCru(CruFastaFile, gene, ProteinSeqFasta_path)\n",
    "            extract_AarBraCru(CruFastaFile, gene, ALL_YABBY_Fasta_path)\n",
    "        elif gene.startswith(\"Sal\") == True:\n",
    "            extract_ChiSalTar(SalFastaFile, gene, ProteinSeqFasta_path)\n",
    "            extract_ChiSalTar(SalFastaFile, gene, ALL_YABBY_Fasta_path)\n",
    "        elif gene.startswith(\"gene\") == True:\n",
    "            extract_ChiSalTar(TarFastaFile, gene, ProteinSeqFasta_path)\n",
    "            extract_ChiSalTar(TarFastaFile, gene, ALL_YABBY_Fasta_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
